# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [88]:
import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv"
df = pd.read_csv(url)
print(df.shape) #10910 rows, 26 Columns.
df.head()

(10910, 26)


Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

In [ ]:
# lets clean all columns names to avoid errors.
df.columns = (
    df.columns
      .str.strip()           # remove leading/trailing spaces
      .str.lower()           # lowercase
      .str.replace(' ', '_') # spaces to underscores
)


In [90]:
df.nunique() # We can see we have a total of 9134 Customers from the 10910 rows, the difference are nan values.

unnamed:_0                       10910
customer                          9134
state                                5
customer_lifetime_value           8041
response                             2
coverage                             3
education                            5
effective_to_date                   59
employmentstatus                     5
gender                               2
income                            5694
location_code                        3
marital_status                       3
monthly_premium_auto               202
months_since_last_claim             36
months_since_policy_inception      100
number_of_open_complaints            6
number_of_policies                   9
policy_type                          3
policy                               9
renew_offer_type                     4
sales_channel                        4
total_claim_amount                5106
vehicle_class                        6
vehicle_size                         3
vehicle_type             

In [91]:
df['response'].unique() 

array(['No', 'Yes', nan], dtype=object)

- We have 3 values in 'response' column, 
  (Yes, No and NaN), this means the column needs no cleaning, there are no typing errors only nan's, 
   we will ignore NaN values for now.

In [92]:
df['response'].isna().sum() 

np.int64(631)

- 631 nan values from the total data frame, this confirms we can ignore nan values.

In [93]:
# New Data Frame:

new_df = df[
    (df['total_claim_amount'] < 1000) & 
    (df['response'] == "Yes")
]

new_df.head()


unnamed:_0 customer       state  customer_lifetime_value response  \
3            3  XL78013      Oregon             22332.439460      Yes   
8            8  FM55990  California              5989.773931      Yes   
15          15  CW49887  California              4626.801093      Yes   
19          19  NJ54277  California              3746.751625      Yes   
27          27  MQ68407      Oregon              4376.363592      Yes   

    coverage education effective_to_date employmentstatus gender  ...  \
3   Extended   College           1/11/11         Employed      M  ...   
8    Premium   College           1/19/11         Employed      M  ...   
15     Basic    Master           1/16/11         Employed      F  ...   
19  Extended   College           2/26/11         Employed      F  ...   
27   Premium  Bachelor           2/28/11         Employed      F  ...   

    number_of_open_complaints number_of_policies     policy_type  \
3                         0.0                  2  Corporate Auto   
8                         0.0                  1   Personal Auto   
15                        0.0                  1    Special Auto   
19                        1.0                  1   Personal Auto   
27                        0.0                  1   Personal Auto   

          policy  renew_offer_type  sales_channel  total_claim_amount  \
3   Corporate L3            Offer2         Branch          484.013411   
8    Personal L1            Offer2         Branch          739.200000   
15    Special L1            Offer2         Branch          547.200000   
19   Personal L2            Offer2    Call Center           19.575683   
27   Personal L3            Offer2          Agent           60.036683   

    vehicle_class vehicle_size vehicle_type  
3   Four-Door Car      Medsize            A  
8      Sports Car      Medsize          NaN  
15            SUV      Medsize          NaN  
19   Two-Door Car        Large            A  
27  Four-Door Car      Medsize          NaN  

[5 rows x 26 columns]

In [94]:
print(new_df.nunique()) 

unnamed:_0                       1399
customer                         1248
state                               5
customer_lifetime_value           208
response                            1
coverage                            3
education                           5
effective_to_date                  57
employmentstatus                    5
gender                              2
income                            176
location_code                       3
marital_status                      3
monthly_premium_auto               69
months_since_last_claim            36
months_since_policy_inception      89
number_of_open_complaints           6
number_of_policies                  9
policy_type                         3
policy                              9
renew_offer_type                    3
sales_channel                       4
total_claim_amount                156
vehicle_class                       4
vehicle_size                        3
vehicle_type                        1
dtype: int64

In [95]:
new_df['response'].value_counts(dropna=False)


response
Yes    1399
Name: count, dtype: int64

- After filtering, from the total 9134 customers, we only have 1248 Clients who said 'Yes'.
- But 'Response' column gives us 1399 'Yes' counts, which means some clients said Yes more then once.

-------------------------------------------------------------------------------------------------------------------

2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

In [96]:
yes_clients = df[df['response'] == 'Yes']


In [ ]:
mean_values = yes_clients.groupby(['policy_type', 'gender'])[
    ['monthly_premium_auto', 'customer_lifetime_value']
].mean()

yes_clients.head()

unnamed:_0 customer       state  customer_lifetime_value response  \
3            3  XL78013      Oregon             22332.439460      Yes   
8            8  FM55990  California              5989.773931      Yes   
15          15  CW49887  California              4626.801093      Yes   
19          19  NJ54277  California              3746.751625      Yes   
27          27  MQ68407      Oregon              4376.363592      Yes   

    coverage education effective_to_date employmentstatus gender  ...  \
3   Extended   College           1/11/11         Employed      M  ...   
8    Premium   College           1/19/11         Employed      M  ...   
15     Basic    Master           1/16/11         Employed      F  ...   
19  Extended   College           2/26/11         Employed      F  ...   
27   Premium  Bachelor           2/28/11         Employed      F  ...   

    number_of_open_complaints number_of_policies     policy_type  \
3                         0.0                  2  Corporate Auto   
8                         0.0                  1   Personal Auto   
15                        0.0                  1    Special Auto   
19                        1.0                  1   Personal Auto   
27                        0.0                  1   Personal Auto   

          policy  renew_offer_type  sales_channel  total_claim_amount  \
3   Corporate L3            Offer2         Branch          484.013411   
8    Personal L1            Offer2         Branch          739.200000   
15    Special L1            Offer2         Branch          547.200000   
19   Personal L2            Offer2    Call Center           19.575683   
27   Personal L3            Offer2          Agent           60.036683   

    vehicle_class vehicle_size vehicle_type  
3   Four-Door Car      Medsize            A  
8      Sports Car      Medsize          NaN  
15            SUV      Medsize          NaN  
19   Two-Door Car        Large            A  
27  Four-Door Car      Medsize          NaN  

[5 rows x 26 columns]

In [ ]:
# Comparing yes clients with total claim amount:

claims_df = (
    yes_clients.groupby(['policy_type', 'gender'])
          .agg(avg_total_claim_amount=('total_claim_amount', 'mean'))
          .reset_index()
          .sort_values('avg_total_claim_amount', ascending=True) # Sorting from the less claim amount to most. 
)
claims_df['avg_total_claim_amount'] = claims_df['avg_total_claim_amount'].round(2) # Only 2 decimals

claims_df


policy_type gender  avg_total_claim_amount
1  Corporate Auto      M                  408.58
5    Special Auto      M                  429.53
0  Corporate Auto      F                  433.74
2   Personal Auto      F                  452.97
4    Special Auto      F                  453.28
3   Personal Auto      M                  457.01

- Corporate Auto: Females have the bigger claim amount.
- Personal Auto: Males have the bigger claim amount.
- Special Auto: Females have the bigger claim amount.
- By gender in general, Males have a total of 1295,12 average claim amount and Females a total of 1339,99 claim amount, which makes males the most profitable and low risk.

------------------------------------------------------------

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [119]:
state_counts = df['state'].value_counts()
state_counts

state
California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: count, dtype: int64

- Since each row represents a customer with a yes response, we have more then 500 customers in every state of this data frame.
-----------------------------------------------------------------------------------------------

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [124]:
clv_stats = df.groupby(['education', 'gender'])['customer_lifetime_value'].agg(['max', 'min', 'median'])

clv_stats = clv_stats.round(2)
clv_stats



max      min   median
education            gender                            
Bachelor             F       73225.96  1904.00  5640.51
                     M       67907.27  1898.01  5548.03
College              F       61850.19  1898.68  5623.61
                     M       61134.68  1918.12  6005.85
Doctor               F       44856.11  2395.57  5332.46
                     M       32677.34  2267.60  5577.67
High School or Below F       55277.45  2144.92  6039.55
                     M       83325.38  1940.98  6286.73
Master               F       51016.07  2417.78  5729.86
                     M       50568.26  2272.31  5579.10

### Education level strongly influences CLV
    - Bachelor degree customers in both genders show the highest maximum CLV values, up to 73225 for females and 67907 for males.
    - High school or below male customers show the single highest CLV values, up to 83325, indicating extremely valuable individuals.
    - After College customers, Doctors are the ones with the lowest max CLV values, indicating these groups do not include high-value outliers.
## Overall conclusion:
### The most valuable segments (based on median CLV) appear to be:
    - High School or Below – Male
    - High School or Below – Female
    - College – Male

### Meanwhile, the least valuable segments (median-wise) are:
    - Doctor : Female
    - Doctor : Male
This suggests that lifetime value is not strictly determined by education level, lower-education groups can still produce highly profitable long-term customers.

------------------------------------------------------------

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [98]:
# your code goes here